scheduler/src/cook/mesos/optimizer.clj

```diff
@@ -19,122 +19,251 @@
             [clj-http.client :as http]
             [clj-time.core :as t]
             [clj-time.coerce :as tc]
-            [clj-time.periodic :as periodic]
-            [clojure.core.async :as async]
             [clojure.tools.logging :as log]
-            [cook.util :refer [lazy-load-var PosNum PosInt NonNegInt]]
+            [cook.util :refer [lazy-load-var NonNegInt PosNum PosInt]]
             [cook.mesos.util :as util]
-            [datomic.api :as d :refer (q)]
-            [schema.core :as s]))
+            [metrics.timers :as timers]
+            [plumbing.core :as pc]
+            [schema.core :as s])
+  (:import (java.util UUID)))
 
 (def TimePeriodMs NonNegInt)
 
+
+;; Elements missing:
+;;   1. Attributes (chip set, az, ..) -- have a combinatorial problem here
+;;   2. Price
+;;   3. Prob preemption in 10 minutes increments (or something like that)
 (def HostInfo
+  "Example:
+   {\"count\" 4
+    \"cpus\" 8
+    \"instance-type\" \"basic\"
+    \"mem\" 240000}"
   {:count NonNegInt
-   :instance-type s/Str
    :cpus PosNum
+   :instance-type s/Str
    :mem PosNum
-   (s/optional-key :gpus) PosNum})
-
-(defprotocol HostFeed
-  "Protocol defining a service to get information on hosts that can be purchased."
-  (get-available-host-info [this] "Returns a list of host info maps conforming
-                                   to HostInfo schema above"))
+   (s/optional-key :attributes) {s/Keyword s/Any}
+   (s/optional-key :gpus) PosNum
+   (s/optional-key :time-to-start) TimePeriodMs})
 
 (def Schedule
-  "Schedule returned by optimizer. Each TimePeriodMs is some milliseconds into the future the optimizer
-   recommendation map applies.
-   The optimizer recommendation map currently only has one key, :suggested-matches which has the matches
-   recommended as a value. The matches recommended are structured as a map from the host type in question
-   to a list of job UUIDs the optimizer recommends scheduling
+  "Schedule returned by optimizer.
```

Comment:
This is a schema? Maybe we should be using [`defschema`](http://plumatic.github.io/schema/schema.core.html#var-defschema) to make that fact more explicit.

Same for the other schema definitions above.